fix(metrics): Enforce metric name length limit [INGEST-1205]

iker-barriocanal opened the pull request:
Discard metrics with too-long names before aggregating them to buckets. A metric name and its tags' keys and values are too long if they exceed the 200-byte threshold.

These limits are configurable in the `config.yml`:

```yml
aggregator:
  max_name_length: 200
  max_tag_key_length: 200
  max_tag_value_length: 200
```

iker-barriocanal committed:
Update changelog

jjbayer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -745,6 +745,19 @@ impl BucketKey {
+    fn validate_name(self, aggregator_config: &AggregatorConfig) -> Result<Self, ()> {
+        if self.metric_name.len() as u64 > aggregator_config.max_name_length {
+            relay_log::error!(
+                "Invalid metric name, too large ({} > max:{}): {:?}",
```
```suggestion
                "Invalid metric name, too long ({} > max:{}): {:?}",
```

jjbayer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -1055,6 +1074,10 @@ impl Aggregator {
 
+        let key = key
+            .validate_name(&self.config)
+            .map_err(|()| AggregateMetricsError)?;
```
Please double check my reasoning, but I think returning an `AggregateMetricsError` here causes the entire batch of buckets to fail: 
https://github.com/getsentry/relay/blob/043479fe570949003964c672ed3dda7253a98292/relay-metrics/src/aggregation.rs#L1134-L1143

Not sure what the best solution is, either swallow the error here or change `merge_all` to swallow the error there.

untitaker replied:
ah right. fwiw that bogus code came from https://github.com/getsentry/relay/pull/1235/files#diff-9bce334fb404ce42d375eab0d36d6bb5d79f4d373c8c92c02f8a70dfea0cc1d8R1083

i'm gonna play around with this a bit and see

jjbayer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -745,6 +745,19 @@ impl BucketKey {
+                aggregator_config.max_name_length,
+                self.metric_name
+            );
+            return Err(());
```
We must apply the same validation to all the strings in `tags` (both key and values).

untitaker commented on relay-metrics/src/aggregation.rs:
```diff
@@ -745,6 +745,46 @@ impl BucketKey {
+    ///
+    /// If the metric name surpasses the threshold, returns an Err. Tags with surpassing key or
+    /// value lengths are removed from the bucket.
+    fn validate_string_lengths(mut self, aggregator_config: &AggregatorConfig) -> Result<Self, ()> {
```
Please move this function to be separate from the type, ideally static method on aggregator. see https://github.com/getsentry/relay/pull/1235

untitaker reviewed (approved):
feel free to merge after fixing the comment

jan-auer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -745,6 +745,46 @@ impl BucketKey {
+    /// value lengths are removed from the bucket.
+    fn validate_string_lengths(mut self, aggregator_config: &AggregatorConfig) -> Result<Self, ()> {
+        if self.metric_name.len() as u64 > aggregator_config.max_name_length {
+            relay_log::error!(
```
Let's not log an error for this. Either swallow this silently, or do at most `debug!`. In production, this can create too many error messages and sentry events.

jan-auer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -785,6 +825,11 @@ pub struct AggregatorConfig {
+    /// The length the name of a metric is allowed to be.
+    ///
+    /// Defaults to `200` characters.
+    pub max_name_length: u64,
```
Since we need this mostly to compare string lengths, I would make this `usize`.

jan-auer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -785,6 +825,11 @@ pub struct AggregatorConfig {
+
+    /// The length the name of a metric is allowed to be.
+    ///
+    /// Defaults to `200` characters.
```
This is not characters, this is bytes.

jan-auer commented on relay-metrics/src/aggregation.rs:
```diff
@@ -745,6 +745,46 @@ impl BucketKey {
+                );
+                return false;
+            }
+            if value.len() as u64 > aggregator_config.max_name_length {
```
Let's have separate configs for at least metric names and tags, ideally 3 configs for name, key, value.

jan-auer reviewed (commented):
Let's rebase this on top of #1235 once that is merged.

iker-barriocanal committed:
Break validate_bucket_key into smaller methods

The more checks a BucketKey has the more overwhelmed the method would
be. Separated into smaller and more concise methods.

The metric name is validated first because failing means dropping the
bucket. Tags are checked afterward.

iker-barriocanal committed:
Add max name length option to AggregatorConfig

This option defines the maximum length of a metric, not the length of
the tags' keys and values.

Since this value is going to be used mostly to compare strings, making
it `usize` instead of any other integer size, for convenience.

iker-barriocanal committed:
Add missing max_name_length option in AggregatorConfig

iker-barriocanal committed:
Return Err when metric name length is too long

iker-barriocanal committed:
Add metric info to Sentry scope when dropping buckets

iker-barriocanal committed:
Add configurable thresholds for tag keys and values lengths

iker-barriocanal committed:
Ensure tag key and value lengths are short

iker-barriocanal committed:
Add test

iker-barriocanal committed:
Log debug on wrong tag keys and values

iker-barriocanal committed:
Fix lint

untitaker reviewed (approved):
perfect!
